lib/functions.php

```diff
@@ -350,3 +354,76 @@ function put(string $path, string $contents): Promise
 {
     return filesystem()->put($path, $contents);
 }
+
+/**
+ * Asynchronously lock a file
+ * Resolves with a callbable that MUST eventually be called in order to release the lock.
+ *
+ * @param string            $file    File to lock
+ * @param bool              $shared  Whether to acquire a shared or exclusive lock (\LOCK_SH or \LOCK_EX, see PHP flock docs)
+ * @param integer           $polling Polling interval for lock in milliseconds
+ * @param CancellationToken $token   Cancellation token
+ *
+ * @return \Amp\Promise Resolves with a callbable that MUST eventually be called in order to release the lock.
+ */
+function lock(string $file, bool $shared, int $polling = 100, CancellationToken $token = null): Promise
+{
+    return call(static function () use ($file, $shared, $polling, $token) {
+        $operation = $shared ? \LOCK_SH : \LOCK_EX;
+        $token = $token ?? new NullCancellationToken;
+        if (!yield exists($file)) {
+            yield \touch($file);
+            StatCache::clear($file);
+        }
+        $operation |= LOCK_NB;
+        $res = \fopen($file, 'c');
```

Comment:
Shouldn't we just open the file here instead of the additional `touch` above?

Reply:
Well technically yes, but then the touching wouldn't be async.

Reply:
Well, the `fopen` isn't async anyway? I guess we'd have to add the function to `Handle` to have it really async?

Reply:
Well, fopen isn't async, but afaik there is no way to obtain a lock using any of the native async libs supported by amphp.
I thought it'd be nice if we could make at least the `touch` part of the `fopen` async.

Reply:
@trowski What's your opinion here?

Reply:
I'm rather uncomfortable introducing anything that can block when it could be rolled into a simple `Task` executed using parallel.